File: LeetCode/Tag/Dynamic Programming/cpp/322.coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp (n+1, vector<int>(amount+1, -1));
        for (int i = 0; i <= n; ++i) dp[i][0] = 0;
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= amount; ++j) {
                if (j < coins[i-1] || dp[i][j-coins[i-1]] == -1)
                    dp[i][j] = dp[i-1][j];
                else if (dp[i-1][j] == -1)
                    dp[i][j] = 1 + dp[i][j-coins[i-1]];
                else
                    dp[i][j] = min(dp[i-1][j], 1 + dp[i][j-coins[i-1]]);
            }
        }
        return dp[n][amount];
    }
};

File: LeetCode/Tag/Dynamic-Programming/js/322.coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    
    function countCoins(amount, memo={}) {
        if(amount in memo){
            return memo[amount];
        }
        if(amount === 0){
            return 0;
        }

        let minCount = Infinity;
        for(const coin of coins){
            const diff = amount - coin;
            if(diff >= 0){
                const result = countCoins(diff,memo);
                minCount = Math.min(1+result, minCount);
            }

        }
        memo[amount] = minCount;
        return minCount;
    }

    const minCoins = countCoins(amount,{});
    return minCoins === Infinity ? -1 : minCoins;
};
